### PR TITLE
endpoint: Dump policy map only when syncing from the controller

### LINKED ISSUE #8095: policy: do not dump PolicyMap state from datapath upon every policy computation
Now that we have the SelectorCache work plumbed, the biggest cost during policy calculation now is plumbing lots of PolicyMap entries into the datapath.
The current approach for “syncing” PolicyMap state is the following:

* Dump the actual state from datapath, and then do comparisons against it
* If the key is in the BPF map but not in ‘desired state’,  it is removed from datapath
* If the key in desired state but not in dump, add it to datapath.

We also have a controller which runs every minute for syncing all PolicyMap state for an endpoint. This runs once per-minute. This could be made optional via a configuration flag, and / or made less frequent (e.g., once every half hour or every hour).

We should separate out the logic for the controller which checks the state in the datapath and diffs that against the desired state (which is costly)  vs. the logic that is performed during plumbing of policy.

Change the logic performed during policy calculation to do the following:

* Diff desired / realized map state in userspace
* Remove entries that are in realized state, but not in desired state
* Add entries that are in desired state, but not in realized state

This would drastically reduce # of syscalls performed during policy calculation, and thus make policy computation much more performant, especially in the case where a lot of pods are on a single node, since the number of concurrent builds is bounded per-node.

### PR DESCRIPTION
**Pull request by jrajahalme**
Avoid dumping the policy map when updating the endpoint policy. This
relies on the realized policy state being up-to-date. The controller
still syncs with a full dump, so if the realized policy state gets out
of date, the bpf state will still be properly fixed.

Avoiding the full dump on policy update reduces the number of syscalls
needed for the policy update.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

Fixes: #8095

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8104)

### CODE REVIEW

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failure: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/12534/

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/bpf.go [hunk 1]
```diff
@@ -902,6 +902,96 @@ func (e *Endpoint) GetBPFValue() (*lxcmap.EndpointInfo, error) {
...
+			policyKeyToPolicyMapKey := policymap.PolicyKey{
+				Identity:         keyToDelete.Identity,
+				DestPort:         keyToDelete.DestPort,
+				Nexthdr:          keyToDelete.Nexthdr,
+				TrafficDirection: keyToDelete.TrafficDirection,
+			}
+
+			err := e.PolicyMap.DeleteKey(policyKeyToPolicyMapKey)
+			if err != nil {
+				// TODO: Maybe we should ignore delete errors
```
Comment (ianvernon) [inline pkg/endpoint/bpf.go:931]: I think that we should definitely keep it as is and log an error - it indicates that something is out of whack with the endpoint lifecycle.

File: pkg/endpoint/bpf.go [hunk 2]
```diff
@@ -974,37 +1064,20 @@ func (e *Endpoint) syncPolicyMap() error {
 			if err != nil {
 				e.getLogger().WithError(err).Errorf("Failed to delete PolicyMap key %s", entry.Key.String())
 				errors = append(errors, err)
-			} else {
```
Comment (ianvernon) [inline pkg/endpoint/bpf.go:977]: Why is this change necessary?

**Review by ianvernon — COMMENTED** [review]
One minor question, apart from that LGTM

**Comment by ianvernon** [comment]
test-me-please

File: pkg/endpoint/bpf.go (hunk 2 above)
Comment (jrajahalme) [reply pkg/endpoint/bpf.go:977]: Strictly not necessary, but delete is not needed when realized policy pointer is the same as the desired policy pointer. We just tested that the key is not present, so there is no point deleting it.

selector cache plumbing changed the policy lifecycle to copy the desired policy pointer to the realized policy pointer when the policy has been realized.

**Comment by jrajahalme** [comment]
New CI build started two hours after I asked for it.

**Review by ianvernon — APPROVED** [review]
